src/cargo/util/toml/mod.rs

```diff
@@ -97,11 +109,69 @@ fn read_manifest_from_str(
+    let manifest: manifest::TomlManifest = match serde_ignored::deserialize(deserializer, |path| {
         let mut key = String::new();
         stringify(&mut key, &path);
         unused.insert(key);
-    })?;
+    }) {
+        Ok(manifest) => manifest,
+        Err(e) => {
+            return if let Some(span) = e.span() {
+                let (line, column) = translate_position(contents.as_bytes(), span.start);
+                let lines = contents.lines();
+                let line_count = lines.clone().count();
+                let mut source = lines.skip(line).next().unwrap_or_default().to_string();
+                // Add back the new line that was stripped by `.lines()`
+                if line_count > line + 1 {
+                    source.push('\n');
+                // If we are trying to highlight past the end of the file, add a
+                // placeholder character to the end of the line.
+                } else if span.end >= source.len() - 1 {
+                    source.push('∅');
+                }
+                // Make sure we don't try to highlight past the end of the line,
+                // but also make sure we are highlighting at least one character
+                let highlight_end = (column + (span.end - span.start))
+                    .min(source.len())
+                    .max(column + 1);
+                // Get the path to the manifest, relative to the cwd
+                let manifest_path = diff_paths(manifest_file, config.cwd())
+                    .unwrap_or(manifest_file.to_path_buf())
+                    .display()
+                    .to_string();
+                let snippet = Snippet {
+                    title: Some(Annotation {
+                        id: None,
+                        label: Some(e.message()),
+                        annotation_type: AnnotationType::Error,
+                    }),
+                    footer: vec![],
+                    slices: vec![Slice {
+                        source: &source,
+                        line_start: line + 1,
+                        origin: Some(manifest_path.as_str()),
+                        annotations: vec![SourceAnnotation {
+                            range: (column, highlight_end),
+                            label: "",
+                            annotation_type: AnnotationType::Error,
+                        }],
+                        fold: false,
+                    }],
+                };
+                let renderer = Renderer::styled()
+                    .error(ERROR)
+                    .warning(WARN)
+                    .info(NOTE)
+                    .note(NOTE)
+                    .help(NOTE)
+                    .line_no(NOTE);
+                writeln!(config.shell().err(), "{}", renderer.render(snippet))?;
+                Err(AlreadyPrintedError::new(e.into()).into())
+            } else {
+                Err(e.into())
+            };
```

Comment:
Try to make structure match the business logic.  In this case, the indentation is for guard statements and early return / let-else are tools that should instead be used
```rust
let Some(span) = e.span() else {
    return Err(e.into());
};
```